Complete sprint 15

Evo-Eclipse opened the pull request:
Core Data Integration & Store Refactoring

- Added Core Data stack via NSPersistentContainer in AppDelegate, with saveContext method for persistent storage
- Created model entities for Tracker, TrackerCategory, and TrackerRecord with correct relationships and deletion rules
- Migrated tracker/category/record management from manual arrays to dedicated store classes (`TrackerStore`, `TrackerCategoryStore`, `TrackerRecordStore`)
- Set up merge policies and enabled automatic context syncing for stability and background/UI coherence

Tracker Creation Form Enhancements

- Added emoji and color selection with collection views, selection logic, and visual feedback
- Refactored form to use UIScrollView and contentView for improved layout and flexibility
- Centralized validation: create button enabled only with all required fields (title, emoji, color, schedule)

Trackers List & Filtering Improvements

- TrackersViewController now fully migrates UI and logic to use TrackerStore delegate and snapshotFiltered for category grouping, scheduling, and search
- Removed redundant tracker/category management methods, delegating data control to stores
- Default category creation and tracker filtering refactored for consistency and maintainability

UI/UX & Codebase Clean-Up

- Enhanced keyboard dismiss experience across forms and category screens with global tap-to-hide gesture
- Standardized navigation flow and back button handling during new tracker creation
- Updated protocol names, delegate properties, and file structure for clarity
- Migrated tracker identifiers to UUID for Core Data compatibility

General Architecture & Configuration

- All major tracker-related logic now relies on Core Data for future extensibility
- Improved code modularity and separation of concerns for maintainable and scalable development
- Small improvements to error handling, input validation, and assets where necessary

Evo-Eclipse committed:
Implement emoji & color selection for Tracker creation

- Added emoji and color collection views to TrackerCreationFormViewController
- Created reusable EmojiCollectionViewCell and ColorCollectionViewCell components
- Added String and UIColor extensions for emoji and color selection data
- Implemented selection/deselection logic with visual feedback
- Integrated with existing form validation

Evo-Eclipse committed:
Refactor createButton activation logic

- Moved button state logic to updateCreateButtonState for clarity and reuse
- Replaced scattered activation checks with centralized method call
- Behavior and UI unchanged

Evo-Eclipse committed:
Add Core Data with model entities

- Implemented Core Data stack with NSPersistentContainer in AppDelegate
- Added saveContext() for managed object operations
- Created three main entities:
  - TrackerCategoryCoreData: For tracker categories
  - TrackerCoreData: For tracker details and relationships
  - TrackerRecordCoreData: For tracking completion records
- Configured relationships and deletion rules
- Added support for binary data storage

Note: Core Data is not yet integrated with app logic

Helvetii951 commented on Tracker/Stores/TrackerStore.swift:
```diff
@@ -0,0 +1,210 @@
+    func createTracker(_ tracker: Tracker, in categoryTitle: String) {
+        container.performBackgroundTask { ctx in
+            // Find or create category
+            let cat = self.fetchOrCreateCategory(with: categoryTitle, in: ctx)
```
Все же лучше дописывать название переменных. Так код более читаемый. `cat` -> `category`.

Helvetii951 commented on Tracker/Views/ColorCollectionViewCell.swift:
```diff
@@ -0,0 +1,41 @@
+    
+    static let reuseIdentifier = "ColorCell"
+    
+    let colorView: UIView = {
```
Раз уж есть функция `configure`, было бы хорошо сделать компонент приватным.

Helvetii951 commented on Tracker/Views/EmojiCollectionViewCell.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    static let reuseIdentifier = "EmojiCell"
+
+    let cardView: UIView = {
+        let view = UIView()
+        view.layer.cornerRadius = 16
+        return view
+    }()
+    
+    let emojiLabel: UILabel = {
+        let label = UILabel()
+        label.font = .systemFont(ofSize: 32)
+        label.textAlignment = .center
+        return label
+    }()
```
Рекомендую сделать приватными и добавить функцию конфигурации ячейки.  Если мы поменяем внутреннюю реализацию UI ячейки, нам не придется переписывать код в контроллере.

Helvetii951 commented on Tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,122 @@
+    func categoryStoreDidChange(sectionChanges: [StoreSectionChange], objectChanges: [StoreObjectChange])
+}
+
+final class TrackerCategoryStore: NSObject {
```
Сторам очень нежелательно знать о UI слое(`UIKit`). Стору нужно знать только то, что ему требуется для хранения данных. Рекомендую убрать упоминание о `UIApplication`, а `NSManagedObjectContext` передавать через инициализатор.

Helvetii951 commented on Tracker/Stores/TrackerCategoryStore.swift:
```diff
@@ -0,0 +1,122 @@
+        self.container = container
+        self.context = container.viewContext
+        super.init()
+        try? fetchedResultsController.performFetch()
```
`try` лучше всегда оборачивать в do-catch. Так мы быстрее сможем отследить ошибку, записав ее в логи или сервисы статистики.

Helvetii951 commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -31,6 +32,33 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+    // MARK: - Core Data Saving support
+
+    func saveContext () {
+        let context = persistentContainer.viewContext
+        if context.hasChanges {
```
Сейчас в приложении используется контекст, связанный с главным потоком.

В CoreData предусмотрена многопоточность:

[Доки Apple](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/CoreData/Concurrency.html)
[Неплохая вводная статья](https://cocoacasts.com/core-data-and-concurrency)

Все хорошо, просто потом почитай.

Helvetii951 commented on Tracker/Application/AppDelegate.swift:
```diff
@@ -31,6 +32,33 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+                try context.save()
+            } catch {
+                let nserror = error as NSError
+                fatalError("Unresolved error \(nserror), \(nserror.userInfo)")
```
`fatalError` лучше никогда не использовать. Может привести к очень неприятным падениям в будущем. Рекомендую использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfaxrilure(_:file:line:)).
В таких случаях пишут ошибки в логи или сервисы статистики и событий.

Evo-Eclipse replied:
In this case, using `fatalError` is justified, since the application cannot function without `AppDelegate`. Using `assertionError` would require implementing fallback mechanisms, which would complicate the code.

Evo-Eclipse committed:
Extract UIKit dependencies from CoreData stores

- Added AppColor struct as UIKit-free color model with Codable support
- Implemented dependency injection via NSPersistentContainer for TrackerStore, CategoryStore, RecordStore
- Removed UIApplication.shared.delegate access from convenience initializers

Evo-Eclipse committed:
Implement Core Data stores

- Added Core Data persistence layer with NSFetchedResultsController-backed stores
- Implemented TrackerStore, TrackerCategoryStore, TrackerRecordStore with delegate protocols
- Added UIColor+Coding & Weekday+Coding helpers for Core Data serialization
- Migrated Tracker/TrackerRecord IDs from UInt to UUID for Core Data compatibility
- Updated all UI components to work with UUID-based tracker IDs
- Implemented dependency injection pattern with AppDelegate-backed convenience inits
- Set Core Data viewContext merge policy for background/UI coherence

Evo-Eclipse committed:
Migrate UI to Core Data stores

- Implemented TrackerStore.snapshotFiltered for date-based schedule filtering and case-insensitive search
- Updated TrackersViewController to use dependency-injected stores for category creation, tracker filtering, and data source
- Added TrackerStoreDelegate conformance for real-time UI refresh on store changes
- Removed local categories array and manual add/remove methods, fully delegating to TrackerStore APIs
- Updated TrackerCell delegate methods to use UUID and store APIs
- Adjusted Core Data model deletion rules for Tracker/TrackerRecord relationships (Cascade on Tracker delete)
